runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,435 @@
+             toRelease_.insert(it.first);
+           }
+         }
+         if (toRelease_.size() > 0)
+           atomicSet(&pendingRelease_, 1);
+         atomicSet(&gcRunning_, 0);
+         shallRunCollector_ = false;
+       }
+     }
+     atomicSet(&terminateCollector_, false);
+  }
+
+  void addWorker(void* worker) {
+    Locker lock(&lock_);
+    currentAliveWorkers_++;
+    if (mainWorker_ == nullptr) mainWorker_ = worker;
+  }
+
+  void removeWorker(void* worker) {
+    Locker lock(&lock_);
+    // When exiting the worker - we shall collect the cyclic garbage here.
+    shallCollectGarbage_ = true;
+    shallRunCollector_ = true;
+    CHECK_CALL(pthread_cond_signal(&cond_), "Cannot signal collector")
+    currentAliveWorkers_--;
+  }
+
+  void addRoot(ObjHeader* obj) {
+    COLLECTOR_LOG("add root %p\n", obj);
+    // TODO: we can only add root when collector is not processing, which looks like a limitation,
+    //  instead we can add elements to the side buffer or have a separate lock for that.
+    Locker lock(&lock_);
+    rootset_.insert(obj);
+  }
+
+  void removeRoot(ObjHeader* obj) {
+    COLLECTOR_LOG("remove root %p\n", obj);
+    // Note that we can only remove root when the collector is not processing.
+    Locker lock(&lock_);
+    toRelease_.erase(obj);
+    rootset_.erase(obj);
+  }
+
+  void mutateRoot(ObjHeader* newValue) {
+    // TODO: consider optimization, when clearing value (setting to null) in atomic reference shall not lead
+    //   to invalidation of the collector analysis state.
+    atomicSet(&mutatedAtomics_, 1);
+  }
+
+  bool checkIfShallCollect() {
+    auto tick = atomicAdd(&currentTick_, 1);
+    if (shallCollectGarbage_) return true;
+    auto delta = tick - atomicGet(&lastTick_);
+    if (delta > 10 || delta < 0) {
+      auto currentTimestampUs = konan::getTimeMicros();
+      if (currentTimestampUs - atomicGet(&lastTimestampUs_) > 10000) {
+        // TODO: maybe take another lock.
+        Locker locker(&lock_);
+        lastTick_ = currentTick_;
+        lastTimestampUs_ = currentTimestampUs;
+        shallCollectGarbage_ = true;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void releasePendingUnlocked(void* worker) {
+    // We are not doing that on the UI thread, as taking lock is slow, unless
+    // it happens on deinit of the collector or if there are no other workers.
+    if ((atomicGet(&pendingRelease_) != 0) && ((worker != mainWorker_) || (currentAliveWorkers_ == 1))) {
+      Locker locker(&lock_);
+      COLLECTOR_LOG("clearing %d release candidates on %p\n", toRelease_.size(), worker);
+      for (auto* it: toRelease_) {
+        COLLECTOR_LOG("clear references in %p\n", it)
+        traverseObjectFields(it, [](ObjHeader** location) {
+          ZeroHeapRef(location);
+        });
+      }
+      toRelease_.clear();
+      atomicSet(&pendingRelease_, 0);
+    }
+  }
+
+  void rendezvouz(void* worker) {
```

Comment:
Have you considered the option to run the code below in specific dedicated thread?
E.g. the GC thread itself (which would then require `::memoryState` initialized).
This could simplify interaction with GC thread and make the entire algorithm more independent (including its guarantees).

Reply:
Yes, but actually I find that initiating GC checks bound to the event processing seems better reflect event-driven program behavior.